File: src/main/java/com/itechart/demojavakotlin/service/impl/MovieServiceImpl.java
package com.itechart.demojavakotlin.service.impl;

import com.itechart.demojavakotlin.entity.MovieEntity;
import com.itechart.demojavakotlin.entity.TicketEntity;
import com.itechart.demojavakotlin.exceptions.EntityNotFoundException;
import com.itechart.demojavakotlin.exceptions.InvalidInputDataException;
import com.itechart.demojavakotlin.exceptions.UnprocessableException;
import com.itechart.demojavakotlin.model.MovieRequest;
import com.itechart.demojavakotlin.model.TicketsRequest;
import com.itechart.demojavakotlin.repository.MovieRepository;
import com.itechart.demojavakotlin.service.MovieService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

import static java.lang.String.format;

@Service
@RequiredArgsConstructor
public class MovieServiceImpl implements MovieService {

    private final MovieRepository movieRepository;

    @Override
    @Transactional(readOnly = true)
    //todo #DONE
    public List<MovieEntity> getAll() {
        return movieRepository.findAll();
    }

    @Override
    @Transactional
    //todo #DONE
    public MovieEntity addMovie(final MovieRequest requestMovie) {
        if (requestMovie.getName().isEmpty() || requestMovie.getDirector().isEmpty()) {
            throw new InvalidInputDataException("Empty name or director fields");
        }
        return movieRepository.saveAndFlush(
                MovieEntity.builder()
                        .name(requestMovie.getName())
                        .director(requestMovie.getDirector())
                        .description(requestMovie.getDescription())
                        .ticketsQuantity(0)
                        .ticketPrice(BigDecimal.valueOf(0))
                        .build());
    }

    @Override
    @Transactional
    public void deleteMovie(final UUID id) {
        //todo #DONE
        if (!movieRepository.existsById(id)){
            throw new EntityNotFoundException(format("Movie with id: %s not found", id));
        }
        movieRepository.deleteById(id);
    }

    @Override
    @Transactional(readOnly = true)
    public MovieEntity getMovieByName(final String movieName) {
        //todo #DONE
        if (movieName == null || movieName.isEmpty()) {
            throw new UnprocessableException("Movie name is null or empty");
        }
        return movieRepository.findByName(movieName).orElseThrow(
                () -> new EntityNotFoundException(format("Movie with name: %s not found", movieName)));
    }

    @Override
    @Transactional
    public void addTicketsToMovie(final TicketsRequest ticketsRequest) {
        //todo #DONE
        final MovieEntity foundMovie = getMovieByName(ticketsRequest.getMovieName());

        final int totalQuantityToUpdate = foundMovie.getTicketsQuantity() + ticketsRequest.getQuantity();

        changeTotalTicketQuantity(totalQuantityToUpdate, foundMovie.getId(), ticketsRequest.getPrice());
    }

    @Override
    @Transactional
    public void removeTicketsFromMovie(final TicketsRequest ticketsRequest) {
        //todo #DONE
        final MovieEntity foundMovie = getMovieByName(ticketsRequest.getMovieName());

        final int totalQuantityToUpdate = foundMovie.getTicketsQuantity() - ticketsRequest.getQuantity();

        changeTotalTicketQuantity(totalQuantityToUpdate, foundMovie.getId(), foundMovie.getTicketPrice());
    }

    @Override
    @Transactional
    //todo #DONE
    public void changeTotalTicketQuantity(final int quantity, final UUID id, final BigDecimal ticketPrice) {
        movieRepository.changeTotalTicketsQuantity(quantity, id, ticketPrice);
    }
}

File: kotlinpath/demojavakotlin/src/main/kotlin/by/itechart/demojavakotlin/service/impl/MovieServiceImpl.kt
package by.itechart.demojavakotlin.service.impl

import by.itechart.demojavakotlin.entity.MovieEntity
import by.itechart.demojavakotlin.exception.EntityNotFoundException
import by.itechart.demojavakotlin.exception.InvalidInputDataException
import by.itechart.demojavakotlin.exception.UnprocessableException
import by.itechart.demojavakotlin.model.MovieRequest
import by.itechart.demojavakotlin.model.TicketRequest
import by.itechart.demojavakotlin.repository.MovieRepository
import by.itechart.demojavakotlin.service.MovieService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.math.BigDecimal
import java.util.*

@Service
class MovieServiceImpl(private val movieRepository: MovieRepository) : MovieService {
    @Transactional(readOnly = true)
    override fun getAll(): List<MovieEntity> = movieRepository.findAll()

    @Transactional
    override fun addMovie(requestMovie: MovieRequest): MovieEntity {
        if (requestMovie.name.isEmpty() || requestMovie.director.isEmpty()) {
            throw InvalidInputDataException("Empty name or director fields")
        }
        return movieRepository.saveAndFlush(requestMovie.toEntity())
    }

    @Transactional
    override fun deleteMovie(id: UUID) {
        takeIf { movieRepository.existsById(id) }?.let {
            movieRepository.deleteById(id)
        } ?: throw EntityNotFoundException("Movie with id: $id is not found")
    }

    override fun getMovieByName(movieName: String): MovieEntity {
        return if (movieName.isEmpty()) {
            throw UnprocessableException("Movie name is empty")
        } else
            checkNotNull(movieRepository.findByName(movieName)) {
                EntityNotFoundException("Movie with name: $movieName if not found")
            }
    }

    @Transactional
    override fun addTicketsToMovie(ticketRequest: TicketRequest) {
        val (_, quantity, _, movieName) = ticketRequest
        val foundMovie = getMovieByName(movieName)
        val totalQuantityToUpdate = foundMovie.ticketsQuantity + quantity
        changeTotalTicketQuantity(totalQuantityToUpdate, foundMovie.id, foundMovie.ticketPrice)
    }

    override fun removeTicketsFromMovie(ticketRequest: TicketRequest) {
        val foundMovie = getMovieByName(ticketRequest.movieName)
        val totalQuantityToUpdate = foundMovie.ticketsQuantity - ticketRequest.quantity
        changeTotalTicketQuantity(totalQuantityToUpdate, foundMovie.id, foundMovie.ticketPrice)
    }

    @Transactional
    override fun changeTotalTicketQuantity(quantity: Int, id: UUID, ticketPrice: BigDecimal) =
        movieRepository.changeTotalTicketsQuantity(quantity, id, ticketPrice)

    private fun MovieRequest.toEntity() =
        MovieEntity(
            name = this.name,
            director = this.director,
            description = this.description
        )

}
